dev/src/index.ts

```diff
@@ -1166,144 +1166,100 @@ export class Firestore {
-          // 'end' event we intend to forward here. We therefore need to wait
-          // until the API consumer registers their listeners (in the .then()
-          // call) before emitting any further events.
-          resolve();
-
-          // We execute the forwarding of the 'end' event via setTimeout() as
-          // V8 guarantees that the above the Promise chain is resolved before
-          // any calls invoked via setTimeout().
-          setTimeout(() => {
-            if (endCalled) {
-              logger(
-                'Firestore._initializeStream',
-                requestTag,
-                'Forwarding stream close'
-              );
-              resultStream.emit('end');
-            }
-          }, 0);
+          logger('Firestore._initializeStream', requestTag, 'Releasing stream');
+          resolve(resultStream);
         }
-      };
-
-      // We capture any errors received and buffer them until the caller has
-      // registered a listener. We register our event handler as early as
-      // possible to avoid the default stream behavior (which is just to log and
-      // continue).
-      resultStream.on('readable', () => {
-        streamReady();
-      });
+      }
 
-      resultStream.on('end', () => {
+      function streamEnded() {
         logger(
           'Firestore._initializeStream',
           requestTag,
           'Received stream end'
         );
-        endCalled = true;
         streamReady();
-      });
+        resultStream.unpipe(backendStream);
+      }
 
-      resultStream.on('error', err => {
-        logger(
-          'Firestore._initializeStream',
-          requestTag,
-          'Received stream error:',
-          err
-        );
-        // If we receive an error before we were able to receive any data,
-        // reject this stream.
+      backendStream.on('data', () => streamReady());
+      backendStream.on('end', () => streamEnded());
+      backendStream.on('close', () => streamEnded());
+
+      backendStream.on('error', err => {
         if (!streamInitialized) {
+          // If we receive an error before we were able to receive any data,
+          // reject this stream.
           logger(
             'Firestore._initializeStream',
             requestTag,
             'Received initial error:',
             err
           );
-          streamInitialized = true;
           reject(err);
         } else {
-          errorReceived = err;
+          logger(
+            'Firestore._initializeStream',
+            requestTag,
+            'Received stream error:',
+            err
+          );
+          // We execute the forwarding of the 'error' event via setTimeout() as
+          // V8 guarantees that the Promise chain returned from this method
+          // is resolved before any code executed via setTimeout(). This allows
+          // the caller to attach an error handler.
+          setTimeout(() => {
```

Comment:
setImmediate?

Reply:
I tried this earlier and it broke my unit tests. It works now though, so I updated this call.